xss_helpers.go

```diff
@@ -38,104 +33,104 @@ func isBlackAttr(s string) int {
 		return attributeTypeNone
 	}
 
+	upperS := strings.ToUpper(strings.ReplaceAll(s, "\x00", ""))
 	if length >= 5 {
 		// javascript on.*
 		if strings.ToUpper(s[:2]) == "ON" {
 			// got javascript on- attribute name
 			return attributeTypeBlack
 		}
 
-		if strings.ToUpper(strings.ReplaceAll(s, "\x00", "")) == "XMLNS" ||
-			strings.ToUpper(strings.ReplaceAll(s, "\x00", "")) == "XLINK" {
+		if upperS == "XMLNS" || upperS == "XLINK" {
 			// got xmlns or xlink tags
 			return attributeTypeBlack
 		}
 	}
 
 	for _, black := range blacks {
-		if strings.ToUpper(strings.ReplaceAll(s, "\x00", "")) == black.name {
+		if upperS == black.name {
 			// got banner attribute name
 			return black.attributeType
 		}
 	}
 	return attributeTypeNone
 }
 
-func htmlDecodeByteAt(s string, consumed *int) int {
+func htmlDecodeByteAt(s string, consumed int) (int, int) {
 	length := len(s)
 	val := 0
 
 	if length == 0 {
-		*consumed = 0
-		return byteEOF
+		consumed = 0
+		return byteEOF, consumed
 	}
 
-	*consumed = 1
+	consumed = 1
 	if s[0] != '&' || length < 2 {
-		return int(s[0])
+		return int(s[0]), consumed
 	}
 
 	if s[1] != '#' {
 		// normally this would be for named entities
 		// but for this case we don't actually care
-		return '&'
+		return '&', consumed
 	}
 
 	if s[2] == 'x' || s[2] == 'X' {
 		ch := int(s[3])
 		ch = gsHexDecodeMap[ch]
 		if ch == 256 {
 			// degenerate case '&#[?]'
-			return '&'
+			return '&', consumed
 		}
 		val = ch
 		i := 4
 
 		for i < length {
 			ch = int(s[i])
 			if ch == ';' {
-				*consumed = i + 1
-				return val
+				consumed = i + 1
+				return val, consumed
 			}
 			ch = gsHexDecodeMap[ch]
 			if ch == 256 {
-				*consumed = i
-				return val
+				consumed = i
+				return val, consumed
 			}
 			val = val*16 + ch
 			if val > 0x1000FF {
-				return '&'
+				return '&', consumed
 			}
 			i++
 		}
-		*consumed = i
+		consumed = i
 	} else {
 		i := 2
 		ch := int(s[i])
 		if ch < '0' || ch > '9' {
-			return '&'
+			return '&', consumed
 		}
 		val = ch - '0'
 		i++
 		for i < length {
 			ch = int(s[i])
 			if ch == ';' {
-				*consumed = i + 1
-				return val
+				consumed = i + 1
+				return val, consumed
 			}
 			if ch < '0' || ch > '9' {
-				*consumed = i
-				return val
+				consumed = i
+				return val, consumed
 			}
 			val = val*10 + (ch - '0')
 			if val > 0x1000FF {
-				return '&'
+				return '&', consumed
 			}
 			i++
 		}
-		*consumed = i
+		consumed = i
 	}
-	return val
+	return val, consumed
```

Comment:
Well this isn't so obvious too but you could return the line by `return val, i` in the branches above